go-controller/pkg/crd/userdefinednetwork/v1/types.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=userdefinednetworks,scope=Namespaced
+// +kubebuilder:singular=userdefinednetwork
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// UserDefinedNetwork describe network request for a Namespace.
+type UserDefinedNetwork struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Spec is immutable"
+	// +kubebuilder:validation:XValidation:rule="has(self.topology) && self.topology == 'Layer3' ? has(self.layer3): !has(self.layer3)", message="spec.layer3 is required when topology is Layer3 and forbidden otherwise"
+	// +kubebuilder:validation:XValidation:rule="has(self.topology) && self.topology == 'Layer2' ? has(self.layer2): !has(self.layer2)", message="spec.layer2 is required when topology is Layer2 and forbidden otherwise"
+	// +kubebuilder:validation:XValidation:rule="has(self.topology) && self.topology == 'LocalNet' ? has(self.localNet): !has(self.localNet)", message="spec.localNet is required when topology is LocalNet and forbidden otherwise"
+	// +required
+	Spec UserDefinedNetworkSpec `json:"spec"`
+	// +optional
+	Status UserDefinedNetworkStatus `json:"status,omitempty"`
+}
+
+// UserDefinedNetworkSpec defines the desired state of UserDefinedNetworkSpec.
+// +union
+type UserDefinedNetworkSpec struct {
+	// topology describes network configuration.
+	// Allowed values are "Layer3", "Layer2", "LocalNet".
+	// Layer3 topology creates one logical switch per node - each with a different subnet - and a router interconnecting all the logical switches.
```

Comment:
*and routers interconnect all of the logical switch subnets.

There is more than just 1 router in IC.

Reply:
right, so I thought that we don't want to confuse uers too much with the internals of ovn-k topology, and that logically it is equivalent to just interconnected node switches. Do you think it is useful to give these extra details here or will it confuse people more?

Reply:
well then to be generic I think you could just say "Layer3 topology creates a layer 2 segment per node, each with a different subnet. Layer 3 routing is used to interconnect node subnets."